server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -687,9 +731,33 @@ private void postCreateSnapshot(Long volumeId, Long snapshotId, Long policyId) {
             }
         }
 
-        if (snapshot != null && snapshot.isRecursive()) {
+        if (snapshot == null) {
+            return;
+        }
+
+        if (snapshot.isRecursive()) {
             postCreateRecurringSnapshotForPolicy(userId, volumeId, snapshotId, policyId);
         }
+
+        if (HypervisorType.KVM.equals(snapshot.getHypervisorType()) && kvmIncrementalSnapshot.valueIn(clusterId)) {
+            endChainIfNeeded(snapshotId);
+        }
+    }
+
+    private void endChainIfNeeded(Long snapshotId) {
+        SnapshotDataStoreVO snapshotDataStoreVo = _snapshotStoreDao.findOneBySnapshotId(snapshotId);
```

Comment:
```suggestion
        SnapshotDataStoreVO snapshotDataStoreVo = _snapshotStoreDao.findOneBySnapshotId(snapshotId, zoneId);
```
if there are copies to different zones you could get as a result one of them